### PR TITLE
Gripper

### PR DESCRIPTION
**Pull request by rakrakon**

### CODE REVIEW

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import com.revrobotics.SparkLimitSwitch
+import edu.wpi.first.units.Units
+import edu.wpi.first.wpilibj.DigitalInput
+import edu.wpi.first.wpilibj.RobotController
+import edu.wpi.first.wpilibj.Timer
+
+class GripperIOReal : GripperIO {
+    override val inputs = LoggedGripperInputs()
+    private val rollerMotor: CANSparkMax =
+        CANSparkMax(0, CANSparkLowLevel.MotorType.kBrushless) // TODO: Replace deviceID with real value
```
Comment (katzuv) [inline src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt:15]: You can determine this now.

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    override fun setRollerMotorPower(power: Double) {
+        rollerMotor.set(power)
+    }
+
+    private fun hasNote(): Boolean = !sensor.get()
+
+    override fun updateInputs() {
+        inputs.rollerMotorVoltage.mut_replace(rollerMotor.get() * RobotController.getBatteryVoltage(), Units.Volts)
```
Comment (katzuv) [inline src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt:39]: [`getBusVoltage()`](https://codedocs.revrobotics.com/java/com/revrobotics/cansparkbase#getBusVoltage()) is perhaps a better way to do that.

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOSim.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package frc.robot.subsystems.gripper
+
+import com.ctre.phoenix6.controls.DutyCycleOut
+import edu.wpi.first.units.Units
+import edu.wpi.first.wpilibj.Timer
+import frc.robot.lib.motors.TalonFXSim
+
+class GripperIOSim : GripperIO {
+    override val inputs = LoggedGripperInputs()
+    private val rollerMotor = TalonFXSim(1, 1.0, 0.5, 1.0)
```
Comment (katzuv) [inline src/main/kotlin/frc/robot/subsystems/gripper/GripperIOSim.kt:10]: If the real IO uses a Spark MAX, there is no reason to simulate it with a Talon FX.

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        inputs.hasNote = hasNote()
```
Comment (katzuv) [inline src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt:40]: Why is the method `hasNote()` needed?
```suggestion
        inputs.hasNote = !sensor.get()
```

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt (hunk 4 above)
Comment (rakrakon) [reply src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt]: d82e44c3e

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt (hunk 2 above)
Comment (rakrakon) [reply src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt]: 3f800336

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt (hunk 1 above)
Comment (rakrakon) [reply src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt]: ed98a5be31f439d0ac3f590d617fae254c91d42a
eb8bc115e1da82b7347fbe2e539f56f2271cc27e

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOSim.kt (hunk 3 above)
Comment (rakrakon) [reply src/main/kotlin/frc/robot/subsystems/gripper/GripperIOSim.kt]: e31422fbd6373b3bbbf2559e9f6083674d91da43

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt (hunk 2 above)
Comment (vichik123) [reply src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt]: > [`getBusVoltage()`](https://codedocs.revrobotics.com/java/com/revrobotics/cansparkbase#getBusVoltage()) is perhaps a better way to do that.

I'm actually not sure that gives what you want. I don't rememeber 100 percent though.

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt (hunk 4 above)
Comment (vichik123) [reply src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt]: I think the `hasNote()` function was more complicated than simply `!sensor.get()`.

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt (hunk 2 above)
Comment (katzuv) [reply src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt]: > Returns: The voltage fed into the motor controller.

Isn't that the desired value?

File: src/main/kotlin/frc/robot/subsystems/gripper/Gripper.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    fun setRollerPower(power: Double): Command {
+        return run {
+            rollerPowerSetPoint = power
+            io.setRollerMotorPower(power)
+        }.withName("Set Roller Power")
+    }
+
+    override fun periodic() {
+        io.updateInputs()
+        if (timer.advanceIfElapsed(0.1)) Logger.processInputs(this::class.simpleName, io.inputs)
```
Comment (katzuv) [inline src/main/kotlin/frc/robot/subsystems/gripper/Gripper.kt:48]: Always wrap condition bodies in `{}`.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/kotlin/frc/robot/subsystems/gripper/Gripper.kt (hunk 5 above)
Comment (rakrakon) [reply src/main/kotlin/frc/robot/subsystems/gripper/Gripper.kt]: 65e573d0f6fe9cfcb7d53247c534546dfbc52805

File: src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt (hunk 4 above)
Comment (rakrakon) [reply src/main/kotlin/frc/robot/subsystems/gripper/GripperIOReal.kt]: The additional logic was to be able to simulate the sensor.

File: src/main/kotlin/frc/robot/subsystems/gripper/Gripper.kt (hunk 5 above)
Comment (Emma03L) [reply src/main/kotlin/frc/robot/subsystems/gripper/Gripper.kt]: Do you really need to process the inputs every 0.02 seconds? It can cause problems.

**Pull request merged**